File: backend/app/Models/Budget.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class Budget extends Model
{
    use HasFactory;

    protected $fillable = [
        'user_id',
        'name',
        'description',
        'amount',
        'period',
        'start_date',
        'end_date',
        'is_active',
        'color',
    ];

    protected $casts = [
        'amount' => 'decimal:2',
        'is_active' => 'boolean',
        'start_date' => 'datetime',
        'end_date' => 'datetime',
        'created_at' => 'datetime',
        'updated_at' => 'datetime',
    ];

    protected $appends = [
        'spent_amount',
        'remaining_amount',
        'percentage_used',
        'status',
    ];

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function categories(): HasMany
    {
        return $this->hasMany(BudgetCategory::class);
    }

    public function alerts(): HasMany
    {
        return $this->hasMany(BudgetAlert::class);
    }

    public function getSpentAmountAttribute(): float
    {
        return (float) $this->categories()
            ->with('category.transactions')
            ->get()
            ->sum(fn ($bc) => $bc->category->transactions()
                ->whereBetween('transaction_date', [
                    $this->start_date,
                    $this->end_date,
                ])
                ->sum('amount'));
    }

    public function getRemainingAmountAttribute(): float
    {
        return max(0, (float) ($this->amount - $this->spent_amount));
    }

    public function getPercentageUsedAttribute(): float
    {
        if ($this->amount == 0) {
            return 0;
        }
        return round(($this->spent_amount / $this->amount) * 100, 2);
    }

    public function getStatusAttribute(): string
    {
        $percentage = $this->percentage_used;
        if ($percentage >= 100) {
            return 'exceeded';
        } elseif ($percentage >= 80) {
            return 'warning';
        } elseif ($percentage >= 50) {
            return 'moderate';
        }
        return 'healthy';
    }

    public function scopeActive($query)
    {
        return $query->where('is_active', true);
    }

    public function scopeForUser($query, int $userId)
    {
        return $query->where('user_id', $userId);
    }

    public function scopeByPeriod($query, string $period)
    {
        return $query->where('period', $period);
    }

    public function scopeCurrent($query)
    {
        return $query->where('start_date', '<=', now())
            ->where('end_date', '>=', now());
    }
}

File: mobile/lib/models/budget.dart
import 'transaction.dart';

class BudgetCategory {
  final int id;
  final int categoryId;
  final double allocatedAmount;
  final TransactionCategory category;

  BudgetCategory({
    required this.id,
    required this.categoryId,
    required this.allocatedAmount,
    required this.category,
  });

  factory BudgetCategory.fromJson(Map<String, dynamic> json) {
    return BudgetCategory(
      id: json['id'] ?? 0,
      categoryId: json['category_id'] ?? 0,
      allocatedAmount: (json['allocated_amount'] ?? 0).toDouble(),
      category: TransactionCategory.fromJson(json['category'] ?? {}),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'category_id': categoryId,
      'allocated_amount': allocatedAmount,
      'category': category.toJson(),
    };
  }
}

class Budget {
  final int id;
  final String name;
  final String? description;
  final double amount;
  final String period;
  final String startDate;
  final String endDate;
  final bool isActive;
  final String? color;
  final double spentAmount;
  final double remainingAmount;
  final double percentageUsed;
  final List<BudgetCategory> categories;

  Budget({
    required this.id,
    required this.name,
    this.description,
    required this.amount,
    required this.period,
    required this.startDate,
    required this.endDate,
    required this.isActive,
    this.color,
    required this.spentAmount,
    required this.remainingAmount,
    required this.percentageUsed,
    required this.categories,
  });

  factory Budget.fromJson(Map<String, dynamic> json) {
    var catsList = json['categories'] as List? ?? [];
    List<BudgetCategory> parsedCats = catsList.map((c) => BudgetCategory.fromJson(c)).toList();

    return Budget(
      id: json['id'] ?? 0,
      name: json['name'] ?? '',
      description: json['description'],
      amount: (json['amount'] ?? 0).toDouble(),
      period: json['period'] ?? 'monthly',
      startDate: json['start_date'] ?? '',
      endDate: json['end_date'] ?? '',
      isActive: json['is_active'] ?? true,
      color: json['color'],
      spentAmount: (json['spent_amount'] ?? 0).toDouble(),
      remainingAmount: (json['remaining_amount'] ?? 0).toDouble(),
      percentageUsed: (json['percentage_used'] ?? 0).toDouble(),
      categories: parsedCats,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'amount': amount,
      'period': period,
      'start_date': startDate,
      'end_date': endDate,
      'is_active': isActive,
      'color': color,
      'spent_amount': spentAmount,
      'remaining_amount': remainingAmount,
      'percentage_used': percentageUsed,
      'categories': categories.map((c) => c.toJson()).toList(),
    };
  }
}
